pkg/lite-apiserver/cache/cache_mgr.go

```diff
@@ -0,0 +1,324 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	jsonserializer "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/runtime/serializer/streaming"
+	"k8s.io/apimachinery/pkg/watch"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/kubernetes/scheme"
+	restclientwatch "k8s.io/client-go/rest/watch"
+	"k8s.io/klog"
+
+	"github.com/superedge/superedge/pkg/lite-apiserver/constant"
+	"github.com/superedge/superedge/pkg/lite-apiserver/storage"
+)
+
+type CacheManager struct {
+	storage storage.Storage
+}
+
+func NewCacheManager(storage storage.Storage) *CacheManager {
+	return &CacheManager{
+		storage: storage,
+	}
+}
+
+func (c CacheManager) Cache(req *http.Request, statusCode int, header http.Header, body io.ReadCloser) error {
+	info, ok := apirequest.RequestInfoFrom(req.Context())
+	if !ok {
+		return fmt.Errorf("parse requestInfo error")
+	}
+
+	userAgent := getUserAgent(req)
+	key := cacheKey(userAgent, info)
+	klog.V(4).Infof("cache for %s", key)
+
+	return c.handleCache(key, info.Verb, statusCode, header, body)
+}
+
+func (c CacheManager) handleCache(key string, verb string, statusCode int, header http.Header, body io.ReadCloser) error {
+	// handle watch with streaming
+	if verb == constant.VerbWatch {
+		return c.cacheWatch(key, body)
+	}
+
+	// handle get/list
+	var buf bytes.Buffer
+	n, err := buf.ReadFrom(body)
+	if err != nil {
+		klog.Errorf("failed to get cache response, %v", err)
+		return err
+	}
+	klog.V(4).Infof("cache %d bytes body from response for %s", n, key)
+
+	header.Del(constant.ContentLength)
+
+	if verb == constant.VerbList {
+		return c.cacheList(key, statusCode, header, buf.Bytes())
+	} else if verb == constant.VerbGet {
+		return c.cacheGet(key, statusCode, header, buf.Bytes())
+	}
+
+	return nil
+}
+
+func (c CacheManager) cacheWatch(key string, body io.ReadCloser) error {
+	decoder := getWatchDecoder(body)
+	accessor := meta.NewAccessor()
+
+	for {
+		eventType, obj, err := decoder.Decode()
+		if err != nil {
+			klog.V(2).Infof("watch for key %s decode end: %v", key, err)
+			return err
+		}
+
+		klog.V(6).Infof("new watch event, type=%s, obj=%v", eventType, obj)
+
+		switch eventType {
+		case watch.Bookmark:
+			klog.Infof("watch bookmark for key %s", key)
+		case watch.Error:
+			klog.Infof("watch error for key %s", key)
+		case watch.Added, watch.Modified, watch.Deleted:
+			listData, err := c.storage.LoadList(key)
```

Comment:
Maybe we could refactor codes in this `case`.

Add `continue` in above two cases to make below codes simpler and cleaner?

Reply:
added continue